File: market/integration-test/java/ru/yandex/market/logistics/lom/utils/TskvUtils.java
package ru.yandex.market.logistics.lom.utils;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import javax.annotation.Nonnull;
import javax.annotation.ParametersAreNonnullByDefault;

import lombok.experimental.UtilityClass;
import org.apache.commons.lang3.tuple.Pair;

@UtilityClass
@SuppressWarnings("HideUtilityClassConstructor")
@ParametersAreNonnullByDefault
public class TskvUtils {
    public static final String ENTITY_TYPES = "entity_types";
    public static final String ENTITY_VALUES = "entity_values";
    public static final String EXTRA_VALUES = "extra_values";
    public static final String EXTRA_KEYS = "extra_keys";
    public static final String CODE_KEY = "code";
    public static final String LEVEL_KEY = "level";
    public static final String PAYLOAD_KEY = "payload";
    private static final Pattern TSKV_PATTERN = Pattern.compile("([a-z_]+)=([^\\t]*)\\t?|$");

    @Nonnull
    public static Map<String, String> tskvLogToMap(String log) {
        Map<String, String> result = new HashMap<>();
        Matcher matcher = TSKV_PATTERN.matcher(log);
        while (matcher.find()) {
            result.put(matcher.group(1), matcher.group(2));
        }
        return result;
    }

    @Nonnull
    public static Map<String, String> tskvGetExtra(Map<String, String> tskvMap) {
        List<String> extraKeys = splitValues(tskvMap, EXTRA_KEYS);
        List<String> extraValues = splitValues(tskvMap, EXTRA_VALUES);
        return zip(extraKeys, extraValues);
    }

    @Nonnull
    public static Map<String, String> tskvGetEntities(Map<String, String> tskvMap) {
        List<String> entityKey = splitValues(tskvMap, ENTITY_TYPES);
        List<String> entityValues = splitValues(tskvMap, ENTITY_VALUES);
        return zip(entityKey, entityValues);
    }

    @Nonnull
    private static List<String> splitValues(Map<String, String> tskvMap, String key) {
        return Arrays.stream(tskvMap.getOrDefault(key, "").split(","))
            .map(String::trim)
            .collect(Collectors.toList());
    }

    @Nonnull
    private static <L, R> Map<L, R> zip(List<L> first, List<R> second) {
        return IntStream.range(0, Math.min(first.size(), second.size()))
            .mapToObj(i -> Pair.of(first.get(i), second.get(i)))
            .collect(Collectors.toMap(Pair::getKey, Pair::getValue));
    }
}

File: market/test/kotlin/ru/yandex/market/logistics/mqm/utils/TskvUtils.kt
package ru.yandex.market.logistics.mqm.utils

private const val ENTITY_TYPES = "entity_types"
private const val ENTITY_VALUES = "entity_values"
private const val EXTRA_VALUES = "extra_values"
private const val EXTRA_KEYS = "extra_keys"
private const val CODE_KEY = "code"
private const val PAYLOAD_KEY = "payload"
private const val LEVEL_KEY = "level"

private fun tskvLogToMap(
    log: String,
) = Regex("""([a-z_]+)=([^\t]*)\t?|$""")
    .findAll(log)
    .iterator()
    .asSequence()
    .associate { it.destructured.component1() to it.destructured.component2() }

fun tskvGetExtra(log: String) =
    tskvLogToMap(log).map {
        it.key to
            it.value.split(",").map { element -> element.trimEnd().trimStart() }
    }.toMap()
        .let {
            it.getOrDefault(EXTRA_KEYS, emptyList()).zip(
                it.getOrDefault(EXTRA_VALUES, emptyList())
            )
        }

fun tskvGetEntity(log: String) =
    tskvLogToMap(log).map {
        it.key to
            it.value.split(",").map { element -> element.trimEnd().trimStart() }
    }.toMap()
        .let {
            it.getOrDefault(ENTITY_TYPES, emptyList()).zip(
                it.getOrDefault(ENTITY_VALUES, emptyList())
            )
        }

fun tskvGetEntities(log: String): List<Pair<String, String>> {
    return tskvLogToMap(log).map {
        it.key to it.value.split(",").map { element -> element.trimEnd().trimStart() }
    }
        .find { pair -> pair.first == ENTITY_VALUES }!!
        .second
        .map { value -> value.split(":")[0] to value.split(":")[1] }
}

fun tskvGetByKey(log: String, key: String) =
    Pair(key, tskvLogToMap(log)[key])

fun tskvGetCode(log: String) = tskvGetByKey(log, CODE_KEY).second

fun tskvGetPayload(log: String) = tskvGetByKey(log, PAYLOAD_KEY).second

fun tskvGetLevel(log: String) = tskvGetByKey(log, LEVEL_KEY).second
